akka-stream/src/main/scala/akka/stream/impl/Throttle.scala

```diff
@@ -42,26 +68,34 @@ private[stream] class Throttle[T](cost: Int,
 
       override def onPush(): Unit = {
         val elem = grab(in)
-        val elementCost = scale(costCalculation(elem))
+        val elementCost = costCalculation(elem)
 
-        if (lastTokens >= elementCost) {
-          lastTokens -= elementCost
+        // Avoid calculating exact token count when we are sure that there are enough known tokens
+        if (recordedTokens >= elementCost) {
+          recordedTokens -= elementCost
           push(out, elem)
         } else {
-          val currentTime = now()
-          val currentTokens = Math.min((currentTime - previousTime) * speed + lastTokens, scaledMaximumBurst)
-          if (currentTokens < elementCost)
+          val tokensArrived = (System.nanoTime() - arrivalOfLastRecordedToken) / nanosBetweenTokens
+          val refreshedTokens = Math.min(recordedTokens + tokensArrived, maximumBurst)
+
+          if (refreshedTokens < elementCost)
             mode match {
               case Shaping ⇒
                 currentElement = Some(elem)
-                val waitTime = (elementCost - currentTokens) / speed
-                previousTime = currentTime + waitTime
-                scheduleOnce(timerName, waitTime.nanos)
+                val waitNanos = (elementCost - refreshedTokens) * nanosBetweenTokens
+                // due to the rounding down in tokensArrived this will track the idealized time when the last
+                // recorded integer token arrived, which in this case will be the time once we have all the
+                // tokens available to emit this element (i.e. this is calculated *ahead* as this arrival time
+                // is currently in the future and will be valid after the timer fires).
+                arrivalOfLastRecordedToken += elementCost * nanosBetweenTokens
+                scheduleOnce(timerName, waitNanos.nanos)
               case Enforcing ⇒ failStage(new RateExceededException("Maximum throttle throughput exceeded"))
             }
           else {
-            lastTokens = currentTokens - elementCost
-            previousTime = currentTime
+            recordedTokens = refreshedTokens - elementCost
+            // Never set this timestamp to currentNanos directly! This tracks the *ideal* time of the arrival of
+            // the last recorded token, not when we actually calculated it (currentNanos).
```

Comment:
I think this is incorrect: this only holds unless the bucket was completely full, at which point we cannot fill it further (not even ideally), so we must set the last recorded time to “now”. With the current solution we effectively switch off the burst limit.